### PR TITLE
Fix mkl-static issue for Windows.

### PR DESCRIPTION
**Pull request by xuhancn**
Background:
We found the pytorch Windows release/2.4 performance regression: https://github.com/pytorch/pytorch/issues/130619

After some debug works, I found the pytorch Windows static mkl build options are wrong: 
image
1. Thread lib is wrong.
2. Miss `openmp` lib and config.
> Debug history: https://github.com/pytorch/pytorch/issues/130619#issuecomment-2226782504 and https://github.com/pytorch/pytorch/issues/130619#issuecomment-2226418611

This PR will fix `mkl-static` build options issue.
image

Reference:
image

https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl-link-line-advisor.html#gs.c6izlg

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: cmake/Modules/FindMKL.cmake [hunk 1]
```diff
@@ -88,25 +88,41 @@ ELSE ("${SIZE_OF_VOIDP}" EQUAL 8)
...
-    SET(mklrtls "tbb")
-  ELSE()
-    SET(mklthreads "mkl_gnu_thread" "mkl_intel_thread")
-    SET(mklrtls "gomp" "iomp5")
-  ENDIF()
-  SET(mklifaces  "intel" "gf")
-ELSE(CMAKE_COMPILER_IS_GNUCC)
+
+IF(WIN32)
   IF ("${MKL_THREADING}" STREQUAL "TBB")
```
Comment (malfet) [inline cmake/Modules/FindMKL.cmake:93]: We got rid of all TBB builds recently, perhaps it's a good opportunity to clean up those as well

File: cmake/Modules/FindMKL.cmake [hunk 2]
```diff
@@ -88,25 +88,41 @@ ELSE ("${SIZE_OF_VOIDP}" EQUAL 8)
...
+  ELSE(CMAKE_COMPILER_IS_GNUCC)
+    IF ("${MKL_THREADING}" STREQUAL "TBB")
+      SET(mklthreads "mkl_tbb_thread")
+      SET(mklrtls "tbb")
+    ELSE()
+      SET(mklthreads "mkl_intel_thread")
+      SET(mklrtls "iomp5" "guide")
+    ENDIF()
+    SET(mklifaces  "intel")
+  ENDIF (CMAKE_COMPILER_IS_GNUCC)
```
Comment (malfet) [inline cmake/Modules/FindMKL.cmake:124]: Can you please explain why one needs section that differentiates between gcc and clang?

File: cmake/Modules/FindMKL.cmake [hunk 3]
```diff
@@ -88,25 +88,41 @@ ELSE ("${SIZE_OF_VOIDP}" EQUAL 8)
...
-  SET(mklifaces  "intel")
-ENDIF (CMAKE_COMPILER_IS_GNUCC)
+    SET(mklifaces  "intel")
+ELSE(WIN32)
+  IF(CMAKE_COMPILER_IS_GNUCC)
+    IF ("${MKL_THREADING}" STREQUAL "TBB")
+      SET(mklthreads "mkl_tbb_thread")
+      SET(mklrtls "tbb")
+    ELSE()
+      SET(mklthreads "mkl_gnu_thread" "mkl_intel_thread")
```
Comment (malfet) [inline cmake/Modules/FindMKL.cmake:111]: It looks like this PR changes preference from Intel's OpenMP library to GNU one. I'm fine with it, but would prefer to do it in separate change

File: cmake/Modules/FindMKL.cmake (hunk 2 above)
Comment (xuhancn) [reply cmake/Modules/FindMKL.cmake:124]: It is original code, I don't known why it. I only add a section for `Win32` and move the origianl code to `else` block.

File: cmake/Modules/FindMKL.cmake (hunk 3 above)
Comment (xuhancn) [reply cmake/Modules/FindMKL.cmake:111]: No, I didn't change any thing for original code. 
image

I just move the original code to `else` block of `Win32`.

File: cmake/Modules/FindMKL.cmake (hunk 1 above)
Comment (xuhancn) [reply cmake/Modules/FindMKL.cmake:93]: I suggest to clean up code in further, this PR is only impact Windows build.
If we can fix the performance regression. Maybe we still have chance to cherry pick to `release/2.4`.

**Review by jgong5 — APPROVED** [review]

File: cmake/Modules/FindMKL.cmake (hunk 3 above)
Comment (atalman) [reply cmake/Modules/FindMKL.cmake:111]: Looks like Github change highlight is sometimes off. Here is the correct change view:
Screenshot 2024-07-15 at 9 22 18 AM

**Review by atalman — APPROVED** [review]
lgtm, lets wait for green signal before merging

**Pull request closed without merging**